verification/certs.go

```diff
@@ -65,8 +170,422 @@ func getTcbInfoForHandle(c DPEClient, handle *ContextHandle) (*ContextHandle, Di
+		return err
+	}
+
+	// Remove unhandled extended key usages reported by x509 but defined in spec
+	if err = removeTcgDiceExtendedKeyUsages(certsToProcess); err != nil {
+		return err
+	}
+
+	// Build verify options
+	var opts *x509.VerifyOptions
+	if opts, err = buildVerifyOptions(certChain); err != nil {
+		return err
+	}
+
+	// Certificate chain validation for leaf
+	chains, err := leafCert.Verify(*opts)
+	if err != nil {
+		// Unable to build certificate chain from leaf to root
+		return fmt.Errorf("error verifying DPE leaf: %s", err.Error())
+	}
+
+	// Log certificate chains linked to leaf
+	if len(chains) != 1 {
+		return fmt.Errorf("unexpected number of cert chains: %d", len(chains))
+	}
+	return nil
+}
+
+// Builds Certificate chain verifier parameters.
+func buildVerifyOptions(certChain []*x509.Certificate) (*x509.VerifyOptions, error) {
+	var err error
+	roots := x509.NewCertPool()
+	intermediates := x509.NewCertPool()
+
+	// Root certificate is expected to be in the beginning of the chain, the rest are expected to be intermediates.
+	roots.AddCert(certChain[0])
+
+	for _, cert := range certChain[1:] {
+		if cert.Subject.String() == cert.Issuer.String() {
+			return nil, fmt.Errorf("found a self-signed certificate in middle of certificate chain returned by GetCertificateChain")
+		}
+		intermediates.AddCert(cert)
+	}
+	opts := x509.VerifyOptions{
+		Roots:         roots,
+		Intermediates: intermediates,
+		CurrentTime:   time.Now().UTC(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	return &opts, err
+}
+
+// Gets KeyUsage bitmap and returns as list of KeyUsage name strings.
+func getKeyUsageNames(keyUsage x509.KeyUsage) []string {
+	keyUsageNames := []string{}
+
+	if keyUsage&x509.KeyUsageDigitalSignature != 0 {
+		keyUsageNames = append(keyUsageNames, "DigitalSignature")
+	}
+
+	if keyUsage&x509.KeyUsageContentCommitment != 0 {
+		keyUsageNames = append(keyUsageNames, "ContentCommitment")
+	}
+
+	if keyUsage&x509.KeyUsageKeyEncipherment != 0 {
+		keyUsageNames = append(keyUsageNames, "KeyEncipherment")
+	}
+
+	if keyUsage&x509.KeyUsageDataEncipherment != 0 {
+		keyUsageNames = append(keyUsageNames, "DataEncipherment")
+	}
+
+	if keyUsage&x509.KeyUsageKeyAgreement != 0 {
+		keyUsageNames = append(keyUsageNames, "KeyAgreement")
+	}
+
+	if keyUsage&x509.KeyUsageCertSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CertSign")
+	}
+
+	if keyUsage&x509.KeyUsageCRLSign != 0 {
+		keyUsageNames = append(keyUsageNames, "CRLSign")
+	}
+
+	if keyUsage&x509.KeyUsageEncipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "EncipherOnly")
+	}
+
+	if keyUsage&x509.KeyUsageDecipherOnly != 0 {
+		keyUsageNames = append(keyUsageNames, "DecipherOnly")
+	}
+
+	return keyUsageNames
+}
+
+// Verifies the TCI_Current and TCI_Cumulative of dice tcb information blocks
+func verifyDiceTcbDigest(tcbInfo DiceTcbInfo, wantCurrentTCI []byte, lastCumulativeTCI []byte) error {
+	var err error
```

Comment:
The helper moved out of extendTCI.go as other command-tests also use this function